### DISCUSSION TITLE
How Do I Access The Native HttpServletRequest?

### DISCUSSION THREAD

**Discussion by ptahchiev** [thymeleaf/thymeleaf#970] [Q&A]
Hello,
I've just migrated to the latest spring-boot 3.x, spring 6.x and thymeleaf 3.1.x and now I cannot migrate my code. 
I have several `AbstractAttributeTagProcessor` and I need to access the native `HttpServletRequest` but that seems to be impossible. 
Is there any reason why the `JakartaServletWebRequest` is package protected?

**Comment by conorheffron** [comment]
@ptahchiev 

In a Spring Boot + Thymeleaf application, you can access the HttpServletRequest in several safe and clean ways, depending on your use case.

---

## 1. Access in Controller (Recommended)
 - You can inject HttpServletRequest directly into your controller method and pass needed data to the Thymeleaf template via the Model (No configuration needed prior to adding object to request mapping).

```
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;

@Controller
public class MyController {

    @GetMapping("/example")
    public String examplePage(HttpServletRequest request, Model model) {
        // Get client IP
        String clientIp = request.getRemoteAddr();

        // Add to model for Thymeleaf
        model.addAttribute("clientIp", clientIp);
        model.addAttribute("userAgent", request.getHeader("User-Agent"));

        return "example"; // Thymeleaf template name (example.html)
    }
}
```
#### Thymeleaf template (example.html):

```
<p>Your IP: <span th:text="${clientIp}"></span></p>
<p>User Agent: <span th:text="${userAgent}"></span></p>
```

---

## 2. Access Directly in Thymeleaf (Not Recommended for Logic)
 - Spring Boot automatically exposes HttpServletRequest as #httpServletRequest in Thymeleaf’s expression utility objects.

```
<p>IP: <span th:text="${#httpServletRequest.remoteAddr}"></span></p>
<p>User Agent: <span th:text="${#httpServletRequest.getHeader('User-Agent')}"></span></p>
```
 - ⚠ Note: This is fine for simple display, but avoid putting business logic in templates.

---

## 3. Using @ModelAttribute for Global Access
 - If you need HttpServletRequest data available in all templates:

```
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ModelAttribute;

@ControllerAdvice
public class GlobalModelAttributes {

    @ModelAttribute("clientIp")
    public String clientIp(HttpServletRequest request) {
        return request.getRemoteAddr();
    }
}
```
 - Now clientIp is available in all Thymeleaf templates:

```
<p th:text="${clientIp}"></p>
```
- For another example, please see `ironoc-db` where version number is injected to all templates/views using the same approach: https://github.com/conorheffron/ironoc-db/blob/main/src/main/java/com/ironoc/db/controller/VersionController.java
## ✅ Best Practice:
 - Use @Controller or @ControllerAdvice to prepare data for the view.
 - Use direct #httpServletRequest access only for very simple, non-critical display values (such as version number, current ID of a logged in user, localized messages or navigation bar text items etc).
